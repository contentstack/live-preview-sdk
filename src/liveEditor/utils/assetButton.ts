/**
 * Removes all existing replace asset buttons from the provided visual editor wrapper element.
 * @param visualEditorWrapper - The visual editor wrapper element to remove replace asset buttons from.
 */
export function removeReplaceAssetButton(
    visualEditorWrapper: HTMLDivElement | null
): void {
    if (!visualEditorWrapper) {
        return;
    }

    const existingReplaceButtons = visualEditorWrapper.getElementsByClassName(
        "visual-editor__replace-button"
    );

    for (const existingReplaceButton of Array.from(existingReplaceButtons)) {
        existingReplaceButton.remove();
    }
}

/**
 * Generates a replace asset button element that can be appended to the DOM.
 * @param targetElement - The target element to attach the button to.
 * @param onClickCallback - The callback function to execute when the button is clicked.
 * @returns The generated HTMLButtonElement.
 */
export function generateReplaceAssetButton(
    targetElement: Element,
    onClickCallback: (event: MouseEvent) => void
): HTMLButtonElement {
    const dimension = targetElement.getBoundingClientRect();
    const replaceButton = document.createElement("button");

    replaceButton.addEventListener("click", onClickCallback);
    replaceButton.classList.add("visual-editor__replace-button");
    replaceButton.setAttribute("data-testid", "visual-editor-replace-asset");
    replaceButton.innerHTML = `Replace Asset`;
    replaceButton.style.top = `${dimension.bottom + window.scrollY - 30}px`;
    replaceButton.style.right = `${window.innerWidth - dimension.right}px`;

    return replaceButton;
}
