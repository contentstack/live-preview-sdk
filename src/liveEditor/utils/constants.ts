export const numericInputRegex = /^-?\d*(\.\d*)?([eE][-+]?\d*)?$/;

export const LIVE_EDITOR_FIELD_TYPE_ATTRIBUTE_KEY = "data-cslp-field-type";

export const LIVE_EDITOR_CHANNEL_ID = "live-editor";

export const LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX = 2;
