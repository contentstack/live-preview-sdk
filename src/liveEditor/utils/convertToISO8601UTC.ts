export function convertToISO8601UTC(
    dateString: string,
    timeString: string
): string {
    // Extract the time and offset
    const time = timeString.slice(0, 8); // '00:00:22'
    const offset = timeString.slice(8); // '+0530'

    // get the only date part of the given date
    const datePart = dateString.split("T")[0];

    // Combine date and time into a single string
    const dateTimeString = `${datePart} ${time}${offset}`;

    // Parse the dateTimeString as a Date object
    const date = new Date(dateTimeString);

    // Convert the date to ISO 8601 format in UTC
    const iso8601UTC = date.toISOString();

    return iso8601UTC;
}
