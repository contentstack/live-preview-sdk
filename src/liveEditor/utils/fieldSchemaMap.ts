import { has, isEqual } from "lodash";
import {
    ISchemaFieldMap,
    ISchemaIndividualFieldMap,
} from "./types/index.types";

import liveEditorPostMessage from "./liveEditorPostMessage";
import { LiveEditorPostMessageEvents } from "./types/postMessage.types";

/**
 * Represents a cache for field schemas. Field schemas are
 * used to easily get the field schema based on the field
 * Cslp.
 */
export class FieldSchemaMap {
    private static fieldSchema: {
        [contentTypeUid: string]: ISchemaIndividualFieldMap;
    } = {};

    /**
     * Retrieves the schema field map for a given content type and field Cslp.
     * @param contentTypeUid - The unique identifier of the content type.
     * @param fieldCslp - The Cslp of the field.
     * @returns The schema field map.
     */
    static async getFieldSchema(
        contentTypeUid: string,
        fieldCslp: string
    ): Promise<ISchemaFieldMap> {
        if (FieldSchemaMap.hasFieldSchema(contentTypeUid, fieldCslp)) {
            return Promise.resolve(
                FieldSchemaMap.fieldSchema[contentTypeUid][fieldCslp]
            );
        }

        const data = await liveEditorPostMessage?.send<{
            fieldSchemaMap: ISchemaIndividualFieldMap;
        }>(LiveEditorPostMessageEvents.GET_FIELD_SCHEMA, {
            contentTypeUid: contentTypeUid,
        });

        if (data?.fieldSchemaMap) {
            FieldSchemaMap.fieldSchema[contentTypeUid] = data.fieldSchemaMap;
        }

        return FieldSchemaMap.fieldSchema[contentTypeUid][fieldCslp];
    }

    static hasFieldSchema(contentTypeUid: string, fieldCslp: string): boolean {
        return has(FieldSchemaMap.fieldSchema, [contentTypeUid, fieldCslp]);
    }

    /**
     * Checks if two field schemas are equal.
     * @param firstFieldSchema - The first field schema to compare.
     * @param secondFieldSchema - The second field schema to compare.
     * @returns True if the field schemas are equal, false otherwise.
     */
    static areFieldSchemaEqual(
        firstFieldSchema: ISchemaFieldMap,
        secondFieldSchema: ISchemaFieldMap
    ): boolean {
        return isEqual(firstFieldSchema, secondFieldSchema);
    }

    /**
     * Sets the field schema for a given content type.
     * @param contentTypeUid The unique identifier of the content type.
     * @param fieldSchemaMap The map of individual field schemas.
     */
    static setFieldSchema(
        contentTypeUid: string,
        fieldSchemaMap: ISchemaIndividualFieldMap
    ): void {
        FieldSchemaMap.fieldSchema[contentTypeUid] = fieldSchemaMap;
    }

    /**
     * Clears the field schema cache.
     */
    static clear(): void {
        FieldSchemaMap.fieldSchema = {};
    }
}
