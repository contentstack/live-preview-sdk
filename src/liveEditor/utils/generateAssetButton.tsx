import { render } from "preact";
import ReplaceAssetButtonComponent from "../components/replaceAssetButton";

/**
 * Removes all existing replace asset buttons from the provided visual editor wrapper element.
 * @param visualEditorWrapper - The visual editor wrapper element to remove replace asset buttons from.
 */
export function removeReplaceAssetButton(
    visualEditorWrapper: HTMLDivElement | null
): void {
    if (!visualEditorWrapper) {
        return;
    }

    const existingReplaceButtons = visualEditorWrapper.getElementsByClassName(
        "visual-editor__replace-button"
    );

    for (const existingReplaceButton of Array.from(existingReplaceButtons)) {
        existingReplaceButton.remove();
    }
}

/**
 * Generates a replace asset button element that can be appended to the DOM.
 * @param targetElement - The target element to attach the button to.
 * @param onClickCallback - The callback function to execute when the button is clicked.
 * @returns The generated HTMLButtonElement.
 */
export function generateReplaceAssetButton(
    targetElement: Element,
    onClickCallback: (event: MouseEvent) => void
): HTMLButtonElement {
   
    const visualEditorWrapper = document.querySelector(".visual-editor__container");
    const wrapper = document.createDocumentFragment();
    render(<ReplaceAssetButtonComponent targetElement={targetElement} onClickCallback={onClickCallback}/>, wrapper);
    
    visualEditorWrapper?.appendChild(wrapper);

    const replaceButton = document.querySelector(".visual-editor__replace-button") as HTMLButtonElement;

    return replaceButton;
}
