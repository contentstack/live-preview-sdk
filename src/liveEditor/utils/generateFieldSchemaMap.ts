import mockData from "../ctmap";
import _ from "lodash";
import { ITraverseSchemaVisitor } from "./types/index.types";

export function generateFieldSchemaMap(ctUID: string): Record<string, any> {
    const pageCT = mockData[ctUID];
    const getFieldSchemaMap: ITraverseSchemaVisitor = {
        fieldMap: {},
        should_visit: (_fieldSchema, _path) => {
            return true;
        },
        visit: function (fieldSchema, path) {
            this.fieldMap[path] = fieldSchema;
            if (fieldSchema.data_type === "link") {
                //handle special key for special fields
                this.fieldMap[`${path}.title`] = fieldSchema;
                this.fieldMap[`${path}.url`] = fieldSchema;
            }
            if (fieldSchema.data_type === "file") {
                this.fieldMap[`${path}.url`] = fieldSchema;
            }
            if (fieldSchema.data_type === "blocks") {
                if (!fieldSchema.blocks) {
                    return;
                }
                fieldSchema.blocks.map((block: any) => {
                    this.fieldMap[`${path}.${block.uid}`] = {
                        ...block,
                        data_type: "block",
                        display_name: block.title,
                    };
                });
            }
        },
    };
    traverseSchema(pageCT.schema, [getFieldSchemaMap]);
    return getFieldSchemaMap.fieldMap;
}

function traverseSchema(
    schema: Record<string, unknown>,
    visitors: Array<ITraverseSchemaVisitor>
): void {
    function genPath(prefix: string, path: string) {
        return _.isEmpty(prefix) ? path : [prefix, path].join(".");
    }

    function traverse(fields: any, path: string) {
        path = path || "";
        for (const element of fields) {
            const field = element;
            const currPath = genPath(path, field.uid);

            visitors.forEach((visitor) => {
                if (visitor.should_visit(field, currPath)) {
                    visitor.visit(field, currPath);
                }
            });

            if (field.data_type === "group") traverse(field.schema, currPath);

            if (
                field.data_type === "global_field" &&
                _.isUndefined(field.schema) === false &&
                _.isEmpty(field.schema) === false
            )
                traverse(field.schema, currPath);
            if (field.data_type === "blocks") {
                field.blocks.forEach(function (block: any) {
                    if (block.schema)
                        traverse(block.schema, currPath + "." + block.uid);
                });
            }
            if (field.data_type === "experience_container") {
                field.variations.forEach(function (variation: any) {
                    if (variation.schema)
                        traverse(
                            variation.schema,
                            currPath + "." + variation.uid
                        );
                });
            }
        }
    }
    traverse(schema, "");
}
