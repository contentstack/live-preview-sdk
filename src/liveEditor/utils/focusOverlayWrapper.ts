import { CslpData } from "../../types/cslp.types";
import { VisualEditorCslpEventDetails } from "../../types/liveEditor.types";
import { extractDetailsFromCslp } from "../../utils/cslpdata";
import {
    DATA_CSLP_ATTR_SELECTOR,
    LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX,
} from "./constants";
import { FieldSchemaMap } from "./fieldSchemaMap";
import { cleanIndividualFieldResidual } from "./handleIndividualFields";
import { caretSVG, deleteSVG, moveLeft, moveRight } from "./icon";
import liveEditorPostMessage from "./liveEditorPostMessage";
import { getChildrenDirection } from "./multipleElementAddButton";
import { LiveEditorPostMessageEvents } from "./types/postMessage.types";

/**
 * Adds a focus overlay to the target element.
 * @param targetElement - The element to add the focus overlay to.
 * @param focusOverlayWrapper - The HTMLDivElement that contains the focus overlay.
 * @returns void
 */
export function addFocusOverlay(
    targetElement: Element,
    focusOverlayWrapper: HTMLDivElement
): void {
    const targetElementDimension = targetElement.getBoundingClientRect();

    focusOverlayWrapper.classList.add("visible");

    const distanceFromTop = targetElementDimension.top + window.scrollY;
    const topOverlayDOM = focusOverlayWrapper.querySelector<HTMLDivElement>(
        ".visual-editor__overlay--top"
    );

    if (topOverlayDOM) {
        topOverlayDOM.style.top = "0";
        topOverlayDOM.style.left = "0";
        topOverlayDOM.style.width = "100%";
        topOverlayDOM.style.height = `calc(${distanceFromTop}px - ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
    }

    const bottomOverlayDOM = focusOverlayWrapper.querySelector<HTMLDivElement>(
        ".visual-editor__overlay--bottom"
    );
    if (bottomOverlayDOM) {
        bottomOverlayDOM.style.top = `calc(${
            targetElementDimension.bottom + window.scrollY
        }px + ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
        bottomOverlayDOM.style.height = `calc(${
            window.document.body.scrollHeight -
            targetElementDimension.bottom -
            window.scrollY
        }px - ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
        bottomOverlayDOM.style.left = "0";
        bottomOverlayDOM.style.width = "100%";
    }

    const leftOverlayDOM = focusOverlayWrapper.querySelector<HTMLDivElement>(
        ".visual-editor__overlay--left"
    );
    if (leftOverlayDOM) {
        leftOverlayDOM.style.left = "0";
        leftOverlayDOM.style.top = `calc(${distanceFromTop}px - ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
        leftOverlayDOM.style.height = `calc(${
            targetElementDimension.height
        }px + ${2 * LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
        leftOverlayDOM.style.width = `calc(${targetElementDimension.left}px - ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
    }

    const rightOverlayDOM = focusOverlayWrapper.querySelector<HTMLDivElement>(
        ".visual-editor__overlay--right"
    );
    if (rightOverlayDOM) {
        rightOverlayDOM.style.left = `calc(${targetElementDimension.right}px + ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
        rightOverlayDOM.style.top = `calc(${distanceFromTop}px - ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
        rightOverlayDOM.style.height = `calc(${
            targetElementDimension.height
        }px + ${2 * LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
        rightOverlayDOM.style.width = `calc(${
            window.innerWidth - targetElementDimension.right
        }px - ${LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX}px)`;
    }

    const outlineDOM = focusOverlayWrapper.querySelector<HTMLDivElement>(
        ".visual-editor__overlay--outline"
    );
    if (outlineDOM) {
        outlineDOM.style.top = `${
            targetElementDimension.top + window.scrollY
        }px`;
        outlineDOM.style.height = `${targetElementDimension.height}px`;
        outlineDOM.style.width = `${targetElementDimension.width}px`;
        outlineDOM.style.left = `${targetElementDimension.left}px`;
    }
}

/**
 * Hides the focus overlay and performs necessary cleanup actions when the user clicks outside of the focused element.
 * @param event - The mouse event that triggered the function.
 * @param elements - An object containing references to the focus overlay wrapper, the previously selected editable DOM element, and the visual editor wrapper.
 */
export function hideFocusOverlay(elements: {
    previousSelectedEditableDOM: Element | null;
    visualEditorWrapper: HTMLDivElement | null;
    visualEditorOverlayWrapper: HTMLDivElement | null;
    focusedToolbar: HTMLDivElement | null;
}): void {
    const {
        previousSelectedEditableDOM,
        visualEditorWrapper,
        visualEditorOverlayWrapper,
        focusedToolbar,
    } = elements;

    if (visualEditorOverlayWrapper) {
        visualEditorOverlayWrapper.classList.remove("visible");

        if (previousSelectedEditableDOM) {
            const pseudoEditableElement = visualEditorWrapper?.querySelector(
                "div.visual-editor__pseudo-editable-element"
            );

            if (
                previousSelectedEditableDOM.hasAttribute("contenteditable") ||
                pseudoEditableElement
            ) {
                const actualEditedElement =
                    pseudoEditableElement ||
                    (previousSelectedEditableDOM as HTMLElement);

                liveEditorPostMessage?.send(
                    LiveEditorPostMessageEvents.UPDATE_FIELD,
                    {
                        data:
                            "innerText" in actualEditedElement
                                ? actualEditedElement.innerText
                                : actualEditedElement.textContent,
                        fieldMetadata: extractDetailsFromCslp(
                            previousSelectedEditableDOM.getAttribute(
                                "data-cslp"
                            ) as string
                        ),
                    }
                );
            }

            cleanIndividualFieldResidual({
                overlayWrapper: visualEditorOverlayWrapper,
                previousSelectedEditableDOM: previousSelectedEditableDOM,
                visualEditorWrapper: visualEditorWrapper,
                focusedToolbar: focusedToolbar,
            });
        }
    }
}
export function appendFocusedToolbar(
    eventDetails: VisualEditorCslpEventDetails,
    focusedToolbarElement: HTMLDivElement
) {
    appendFieldPathDropdown(eventDetails, focusedToolbarElement);
    appendMultipleFieldToolbar(eventDetails, focusedToolbarElement);
}
function closeOverlay() {
    document
        .querySelector<HTMLDivElement>(".visual-editor__overlay--top")
        ?.click();
}
export function appendMultipleFieldToolbar(
    eventDetails: VisualEditorCslpEventDetails,
    focusedToolbarElement: HTMLDivElement
) {
    const { editableElement: targetElement, fieldMetadata } = eventDetails;
    FieldSchemaMap.getFieldSchema(
        fieldMetadata.content_type_uid,
        fieldMetadata.fieldPath
    ).then((fieldSchema) => {
        //@ts-ignore
        if (fieldSchema?.multiple) {
            const multipleFieldToolbar = document.createElement("div");
            multipleFieldToolbar.classList.add(
                "visual-editor__focused-toolbar__multiple-field-toolbar"
            );

            const buttonGroup = document.createElement("div");
            buttonGroup.classList.add(
                "visual-editor__focused-toolbar__button-group"
            );

            const deleteSVGButton = document.createElement("button");
            deleteSVGButton.classList.add(
                "visual-editor__button",
                "visual-editor__button--secondary"
            );
            deleteSVGButton.innerHTML = deleteSVG;
            deleteSVGButton.addEventListener("click", (e) => {
                e.preventDefault();
                e.stopPropagation();
                handleDeleteInstance(fieldMetadata);
            });

            const parentPath =
                fieldMetadata?.multipleFieldMetadata?.parentDetails
                    ?.parentCslpValue || "";

            const addDirection = getChildrenDirection(
                targetElement,
                parentPath
            );

            const movePreviousButton = document.createElement("button");
            movePreviousButton.classList.add(
                "visual-editor__button",
                "visual-editor__button--secondary"
            );
            movePreviousButton.innerHTML = moveLeft;
            movePreviousButton.addEventListener("click", (e) => {
                e.preventDefault();
                e.stopPropagation();
                handleMoveInstance(fieldMetadata, "previous");
            });

            const moveNextButton = document.createElement("button");
            moveNextButton.classList.add(
                "visual-editor__button",
                "visual-editor__button--secondary"
            );
            moveNextButton.innerHTML = moveRight;
            moveNextButton.addEventListener("click", (e) => {
                e.preventDefault();
                e.stopPropagation();
                handleMoveInstance(fieldMetadata, "next");
            });

            if (addDirection === "vertical") {
                movePreviousButton.classList.add("visual-editor__rotate--90");
                moveNextButton.classList.add("visual-editor__rotate--90");
            }

            buttonGroup.append(
                movePreviousButton,
                moveNextButton,
                deleteSVGButton
            );
            multipleFieldToolbar.append(buttonGroup);

            focusedToolbarElement.append(multipleFieldToolbar);
        }
    });
}
export function appendFieldPathDropdown(
    eventDetails: VisualEditorCslpEventDetails,
    focusedToolbarElement: HTMLDivElement
) {
    const { editableElement: targetElement, fieldMetadata } = eventDetails;
    const targetElementDimension = targetElement.getBoundingClientRect();

    const distanceFromTop =
        targetElementDimension.top +
        window.scrollY -
        LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX -
        5;
    const distanceFromLeft =
        targetElementDimension.left - LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX;

    focusedToolbarElement.style.top = `${distanceFromTop}px`;
    focusedToolbarElement.style.left = `${distanceFromLeft}px`;
    const parentPaths = collectParentCSLPPaths(targetElement, 2);

    const FieldLabelWrapper = document.createElement("div");
    FieldLabelWrapper.classList.add(
        "visual-editor__focused-toolbar__field-label-wrapper"
    );

    const currentFieldItem = document.createElement("button");
    currentFieldItem.classList.add(
        "visual-editor__focused-toolbar__field-label-wrapper__current-field",
        "visual-editor__button",
        "visual-editor__button--primary"
    );

    const textDiv = document.createElement("div");
    textDiv.classList.add("visual-editor__focused-toolbar__text");
    textDiv.innerText =
        fieldMetadata.fieldPath[fieldMetadata.fieldPath.length - 1];

    const caretIcon = document.createElement("div");
    caretIcon.classList.add(
        "visual-editor__focused-toolbar__field-label-wrapper__caret"
    );
    caretIcon.innerHTML = caretSVG;

    currentFieldItem.append(textDiv, caretIcon);
    focusedToolbarElement.addEventListener("click", (e) => {
        e.preventDefault();

        if (
            (e.target as Element).classList.contains(
                "visual-editor__focused-toolbar__field-label-wrapper__parent-field"
            )
        ) {
            const cslp = (e.target as Element).getAttribute(
                "data-target-cslp"
            ) as string;
            const parentElement = targetElement.closest(
                `[${DATA_CSLP_ATTR_SELECTOR}="${cslp}"]`
            ) as HTMLElement;
            if (parentElement) {
                parentElement.click();
            }
            return;
        }

        if (FieldLabelWrapper.classList.contains("field-label-dropdown-open")) {
            FieldLabelWrapper.classList.remove("field-label-dropdown-open");
            return;
        }
        FieldLabelWrapper.classList.add("field-label-dropdown-open");
    });

    FieldSchemaMap.getFieldSchema(
        fieldMetadata.content_type_uid,
        fieldMetadata.fieldPath
    ).then((fieldSchema) => {
        textDiv.innerText = fieldSchema.display_name;
    });

    FieldLabelWrapper.appendChild(currentFieldItem);

    parentPaths.forEach((path, index) => {
        const parentFieldItem = document.createElement("button");
        parentFieldItem.classList.add(
            "visual-editor__focused-toolbar__field-label-wrapper__parent-field",
            "visual-editor__button",
            "visual-editor__button--secondary",
            "visual-editor__focused-toolbar__text"
        );

        parentFieldItem.setAttribute("data-target-cslp", path);
        parentFieldItem.style.top = `-${(index + 1) * 30}px`;
        const { content_type_uid, fieldPath } = extractDetailsFromCslp(path);
        parentFieldItem.innerText = fieldPath[fieldPath.length - 1];
        FieldLabelWrapper.appendChild(parentFieldItem);
        FieldSchemaMap.getFieldSchema(content_type_uid, fieldPath).then(
            (fieldSchema) => {
                parentFieldItem.innerText = fieldSchema.display_name;
                return;
            }
        );
    });

    focusedToolbarElement.appendChild(FieldLabelWrapper);
}

function handleDeleteInstance(fieldMetadata: CslpData) {
    liveEditorPostMessage
        ?.send(LiveEditorPostMessageEvents.DELETE_INSTANCE, {
            data:
                fieldMetadata.fieldPathWithIndex +
                "." +
                fieldMetadata.multipleFieldMetadata.index,
            fieldMetadata: fieldMetadata,
        })
        .finally(closeOverlay);
}
function handleMoveInstance(
    fieldMetadata: CslpData,
    direction: "previous" | "next"
) {
    //TODO: Disable first and last instance move
    liveEditorPostMessage
        ?.send(LiveEditorPostMessageEvents.MOVE_INSTANCE, {
            data:
                fieldMetadata.fieldPathWithIndex +
                "." +
                fieldMetadata.multipleFieldMetadata.index,
            direction: direction,
            fieldMetadata: fieldMetadata,
        })
        .finally(closeOverlay);
}

function collectParentCSLPPaths(
    targetElement: Element,
    count: number
): Array<string> {
    const cslpPaths: Array<string> = [];
    let currentElement = targetElement.parentElement;

    while (count > 0 || currentElement === window.document.body) {
        if (!currentElement) {
            return cslpPaths;
        }

        if (currentElement.hasAttribute(DATA_CSLP_ATTR_SELECTOR)) {
            cslpPaths.push(
                currentElement.getAttribute(DATA_CSLP_ATTR_SELECTOR) as string
            );
            count--;
        }
        currentElement = currentElement.parentElement;
    }

    return cslpPaths;
}
