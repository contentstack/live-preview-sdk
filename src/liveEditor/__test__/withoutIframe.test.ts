import crypto from "crypto";

import { sleep } from "../../__test__/utils";
import { IConfig } from "../../types/types";
import { getDefaultConfig } from "../../utils/defaults";
import { VisualEditor } from "../index";

Object.defineProperty(globalThis, "crypto", {
    value: {
        getRandomValues: (arr: Array<any>) => crypto.randomBytes(arr.length),
    },
});

describe("When outside the Visual editor, the Visual Editor", () => {
    const mockedConsoleError = jest
        .spyOn(console, "error")
        .mockImplementation((...args) => {
            if (
                args.at(-1) !==
                `contentstack-adv-post-message: No request listener found for event "init"`
            ) {
                console.error(...args);
            }
        });

    afterAll(() => {
        mockedConsoleError.mockRestore();
    });
    test("should have the start editing button", () => {
        const config: IConfig = getDefaultConfig();
        new VisualEditor(config);
        const startEditingButton = document.querySelector(
            `[data-testid="vcms-start-editing-btn"]`
        );
        expect(startEditingButton).toBeTruthy();
    });

    test("should not have clickable elements", async () => {
        const h1 = document.createElement("h1");

        h1.setAttribute(
            "data-cslp",
            "all_fields.blt58a50b4cebae75c5.en-us.single_line"
        );

        document.body.appendChild(h1);

        const config: IConfig = getDefaultConfig();

        new VisualEditor(config);

        await sleep(0);
        h1.click();

        expect(h1).toMatchSnapshot();
        expect(h1.getAttribute("contenteditable")).toBe(null);
    });
});
