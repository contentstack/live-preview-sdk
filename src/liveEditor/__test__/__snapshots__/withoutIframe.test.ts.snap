// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When outside the Visual editor, the Visual Editor should not have clickable elements 1`] = `
<h1
  data-cslp="all_fields.blt58a50b4cebae75c5.en-us.single_line"
/>
`;
